ENH: add inplace cases to fast ufunc loop macros

juliantaylor opened the pull request:
Both gcc and clang don't automatically specialize the inplace case, so
add extra conditions to the loop macros to get the compilers to emit
decent code.
Without them inplace code ends up much slower than the out of place
code.

juliantaylor committed:
ENH: add inplace cases to fast ufunc loop macros

Both gcc and clang don't automatically specialize the inplace case, so
add extra conditions to the loop macros to get the compilers to emit
decent code.
Without them inplace code ends up much slower than the out of place
code.

mhvk commented:
I always wondered why inplace didn't help speed much. Great to see it addressed (even if I cannot really comment on the code).

seberg commented:
Heh, the code is "just the same" with some extra branches that all do the same thing but help the compiler create an optimized version of the branch. Though we should check quite carefully since it might stretch the tests... 

Do you want to create a release note?

seberg commented on numpy/core/src/umath/loops.c.src:
```diff
@@ -128,40 +131,52 @@
+        tout * out = (tout *)op1; \
+        op; \
+    }
+#define BASE_BINARY_LOOP_S(tin, tout, cin, cinp, vin, vinp, op) \
```
Jeesh, this macro argument passing is starting to give me the creeps ;). By now I think its correct, but could maybe use a comment here to explain things a bit more in depth, I am not used to such macros, but took me a bit to figure out that `in1` always ends up pointing to the correct `args[0]`, etc.

seberg replied:
Or maybe it is just one of those things you have to stare at until it makes sense...

seberg commented:
LGTM in general. Do we have basic tests which exercise all of these branches in one way or another? The only thing that worried me a little for a bit was somehow ending up with in1 and in2 swapped (which would not show up in all of the ufuncs).

juliantaylor commented:
the macro is quite ugly, I though about just keeping the original code, but the macro makes clear that the code in all branches is exactly the same.
I don't know if we have tests explicitly test these cases. I should probably check.

juliantaylor commented:
I don't think it needs release notes, these are only used for integer and bitwise ops, the latter I doubt are written inplace often. float loops where already fine as they are manually optimized (good compilers are probably common enough now (as we dropped 2.6 and thus redhat6) that we can also autovectorize those)

but this becomes very significant if the temporary elision branch is merged, that will turn a lot more operations to inplace operations.

juliantaylor commented:
added explicit tests via the code used to test the float loops too, should cover more than necessary (assuming non-buggy compilers)

juliantaylor committed:
TST: fix inplace case of alignment data generator

Due to the lambdas used it didn't actually generate inplace cases.
Fix a few tests that now break due to assuming that the cases are always
out of place.
Also update some numbers to make it more likely to find issues like
loading from wrong array.

juliantaylor commented:
turns out the inplace case of the alignment data generator was broken all along, should be fixed now, luckily everything works :)

seberg commented:
@juliantaylor from the first look, I am not quite sure what that test fix changed?

juliantaylor commented:
I don't know what I was thinking when I original wrote this (its been quite a while) but it uses a lambda to create data and the inplace one does `inp1()  inp1()` which creates two arrays which is not inplace
so I switched to properly pass the same array, see also the commit message

seberg commented:
Sorry, silly me, I only looked at the first two thirds, and the interesting stuff is in the last bit ;).

seberg commented:
Anyway, can't find anything weird, so will merge tomorrow unless something comes up.

seberg commented:
Thanks Julian!

seberg commented:
@juliantaylor, it seems airspeed velocity thinks that this commit made the performance of reduction along the fast axis considerably slower. Any idea why that might be?

pv commented:
The benchmarks are run on Intel Atom + Debian jessie (gcc 4.9.2) if that helps.

seberg commented:
Hmmm, I guess it is not unlikely that newer compilers might not show the slowdown, so likely nothing to worry about (but maybe worth a quick check).

juliantaylor commented:
which benchmark exactly?
the changed loops are not used for the reductions so it shouldn't have changed anything.

pv commented:
See https://pv.github.io/numpy-bench/index.html#regressions?sort=1&dir=desc or https://pv.github.io/numpy-bench/regressions.xml

seberg commented:
Seems like this one with axis=0 and integer loops, which is of course the slow axis, sorry, so it makes a little more sense (since it should go into this branch).
```
class AddReduceSeparate(Benchmark):
    params = [[0, 1], TYPES1]
    param_names = ['axis', 'type']

    def setup(self, axis, typename):
        self.a = get_squares()[typename]

    def time_reduce(self, axis, typename):
        np.add.reduce(self.a, axis=axis)
```

Frankly, I think it should be exactly the case that would get faster, because as far as I can think right now, the thing that supposedly got slower, is the fully contiguous case with `args[2] == args[0]`....

juliantaylor commented:
an atom is a terrible cpu performance wise, quite possible vectorizing just makes things slower there
I can test that on my own next week

seberg commented:
Sounds like its not worth to think about it long...

pv commented:
fwiw, these cases are faster on my intel-core laptop, goes opposite way to atom.
Probably not worth it to optimize for Atom.
